src/monitoring/index.ts

```diff
@@ -0,0 +1,295 @@
+                domain,
+                type: IsomerHostedDomainType.REDIRECTION,
+              } as IsomerHostedDomain)
+          )
+      )
+  }
+
+  /**
+   * This is in charge of fetching all the domains that are are under Isomer, inclusive
+   * of any subdomains and redirects.
+   */
+  getAllDomains() {
+    this.monitoringServiceLogger.info("Fetching all domains")
+    return ResultAsync.combine([
+      this.getAmplifyDeployments().mapErr(
+        (error) => new MonitoringError(error.message)
+      ),
+      this.getRedirectionDomains(),
+      this.getKeyCdnDomains(),
+    ]).andThen(([amplifyDeployments, redirectionDomains, keyCdnDomains]) => {
+      this.monitoringServiceLogger.info("Fetched all domains")
+      return okAsync(
+        [...amplifyDeployments, ...redirectionDomains, ...keyCdnDomains].sort(
+          (a, b) => {
+            const domainA = a.domain
+            const domainB = b.domain
+            if (
+              domainA.startsWith("www.") &&
+              domainA.slice(`www.`.length) === domainB
+            ) {
+              return 0
+            }
+            if (
+              domainB.startsWith("www.") &&
+              domainA === domainB.slice(`www.`.length)
+            ) {
+              return 0
+            }
+            if (domainA === domainB) return 0
+            return domainA > domainB ? 1 : -1
+          }
+        )
+      )
+    })
+  }
+
+  // todo: once /siteup logic is merged into dev, we can add that as to alert isomer team
+  generateReportCard(domains: IsomerHostedDomain[]) {
+    const reportCard: ReportCard[] = []
+
+    const domainResolvers = domains.map(({ domain, type }) => {
+      const aRecord = ResultAsync.fromPromise(
+        dns.resolve(domain, "A"),
+        (e) => e
+      ).orElse(() => okAsync([]))
+      const quadArecord = ResultAsync.fromPromise(
+        dns.resolve(domain, "AAAA"),
+        (e) => e
+      ).orElse(() => okAsync([]))
+
+      const cNameRecord = ResultAsync.fromPromise(
+        dns.resolve(domain, "CNAME"),
+        (e) => e
+      ).orElse(() => okAsync([]))
+
+      const caaRecord = ResultAsync.fromPromise(
+        dns.resolve(domain, "CAA"),
+        (e) => e
+      )
+        .orElse(() => okAsync([]))
+        .map((records) => records.map((record) => record.toString()))
+
+      return ResultAsync.combineWithAllErrors([
+        aRecord,
+        quadArecord,
+        cNameRecord,
+        caaRecord,
+      ])
+        .andThen((resolvedDns) =>
+          okAsync<ReportCard>({
+            domain,
+            type,
+            aRecord: resolvedDns[0],
+            quadArecord: resolvedDns[1],
+            cNameRecord: resolvedDns[2],
+            caaRecord: resolvedDns[3],
+          })
+        )
+        .map((value) =>
+          reportCard.push({
+            ...value,
+          })
+        )
+        .andThen(() => okAsync(reportCard))
+    })
+
+    return ResultAsync.combineWithAllErrors(domainResolvers)
```

Comment:
actually inspecting the signature for `domainResolvers`, the error type is `never`. 

this is because we've discarded the error earlier (`.orElse(() => okAsync([]))`). this seems to suggest that either
1. this won't error out or that
2. all errors are equivalent to an empty response?

wondering if this assumption is true and if it's not, whether there's value in actually logging the error rather than just giving up and returning empty arrays

Reply:
we go with the second assumption. We are later intending to map this over /siteup for correctness, this is just a from of reporting for some level of sanity check that we have x number of domains under our control and they resolve to these dns values

Reply:
not sure i'm understanding you correctly here! if the intent is as a form of reporting, i think preserving the errors and formatting them -> displaying has value. discarding the errors seems to be counter to the intent of reporting?

Reply:
hmm we are doing discarding any errors here. we are simply reporting that we could not resolve to anything by returning an empty string. this is like a dns dig on the terminal and getting that it resolves to nothing mah

Reply:
could i just check how this feature is meant to be used? in my head, the flow is something like this: 
- user queries `/siteup <domain>` 
- we hit the backend here 
- we query for all the records
- if any error 
  - discard error and return []
- if no errors
  - return the resolved records?

i think it's also worth nothing here that the code structure adopted here doesn't quite fit in with `neverthrow` - we're meant to recover and fix errors (you can observe this through the example for `combineWithAllErrors` [here](https://github.com/supermacro/neverthrow?tab=readme-ov-file#resultasynccombinewithallerrors-static-class-method)). this method actually discards successful results and returns **only** the errors.

i don't think there's any actionable wrt updating this to be idiomatic `neverthrow` but it's something we should consider

Reply:
> user queries /siteup <domain>
we hit the backend here
we query for all the records
if any error
discard error and return []
if no errors
return the resolved records?


nope, siteup functionality is to remain as is. once we get a list of domains, the idea is to map through the list of domains via the site up so that every morning, we know if anything is not as what we expect it to be.

all this does is gets a list of domains -> then dig on them. note that the figuring out if the values returned form the dig are errors is not in this pr as of yet. again, the reason why we are ignoring this is because node:dig throws an error if it resolves to nothing, which does not mean that it is an error. it is valid, and expected for for eg isomer.gov.sg to not resolve anything for CAA results.